File: android/src/main/kotlin/com/snowplowanalytics/snowplow_flutter_tracker/SnowplowFlutterTrackerController.kt
// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_flutter_tracker

import android.content.Context

import com.snowplowanalytics.snowplow.Snowplow;
import com.snowplowanalytics.snowplow.configuration.Configuration;
import com.snowplowanalytics.snowplow_flutter_tracker.readers.configurations.DefaultTrackerConfiguration
import com.snowplowanalytics.snowplow_flutter_tracker.readers.messages.*

object SnowplowFlutterTrackerController {

    fun createTracker(values: Map<String, Any>, context: Context) {
        val messageReader = CreateTrackerMessageReader(values)
        val controllers: MutableList<Configuration> = mutableListOf()

        val networkConfigReader = messageReader.networkConfig;
        val networkConfiguration = networkConfigReader.toConfiguration()

        val trackerConfigReader = messageReader.trackerConfig
        if (trackerConfigReader == null) {
            controllers.add(DefaultTrackerConfiguration.toConfiguration(null, context))
        } else {
            controllers.add(trackerConfigReader.toConfiguration(context))
        }

        val subjectConfigReader = messageReader.subjectConfig
        subjectConfigReader?.let { controllers.add(it.toConfiguration()) }

        val gdprConfigReader = messageReader.gdprConfig
        gdprConfigReader?.let { controllers.add(it.toConfiguration()) }

        Snowplow.createTracker(
                context,
                messageReader.namespace,
                networkConfiguration,
                *controllers.toTypedArray()
        )
    }

    fun trackStructured(values: Map<String, Any>) {
        val eventReader = EventMessageReader(values)
        val trackerController = Snowplow.getTracker(eventReader.tracker)
        val structured = eventReader.toStructuredWithContexts()

        trackerController?.track(structured)
    }

    fun trackSelfDescribing(values: Map<String, Any>) {
        val eventReader = EventMessageReader(values)
        val trackerController = Snowplow.getTracker(eventReader.tracker)
        val selfDescribing = eventReader.toSelfDescribingWithContexts()

        trackerController?.track(selfDescribing)
    }

    fun trackScreenView(values: Map<String, Any>) {
        val eventReader = EventMessageReader(values)
        val trackerController = Snowplow.getTracker(eventReader.tracker)
        val screenView = eventReader.toScreenViewWithContexts()

        trackerController?.track(screenView)
    }

    fun trackTiming(values: Map<String, Any>) {
        val eventReader = EventMessageReader(values)
        val trackerController = Snowplow.getTracker(eventReader.tracker)
        val timing = eventReader.toTimingWithContexts()

        trackerController?.track(timing)
    }

    fun trackConsentGranted(values: Map<String, Any>) {
        val eventReader = EventMessageReader(values)
        val trackerController = Snowplow.getTracker(eventReader.tracker)
        val consentGranted = eventReader.toConsentGrantedWithContexts()

        trackerController?.track(consentGranted)
    }

    fun trackConsentWithdrawn(values: Map<String, Any>) {
        val eventReader = EventMessageReader(values)
        val trackerController = Snowplow.getTracker(eventReader.tracker)
        val consentWithdrawn = eventReader.toConsentWithdrawnWithContexts()

        trackerController?.track(consentWithdrawn)
    }

    fun setUserId(values: Map<String, Any>) {
        val messageReader = SetUserIdMessageReader(values)
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        trackerController?.subject?.userId = messageReader.userId
    }

    fun getSessionUserId(values: Map<String, Any>): String? {
        val messageReader = GetParameterMessageReader(values)
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        return trackerController?.session?.userId
    }

    fun getSessionId(values: Map<String, Any>): String? {
        val messageReader = GetParameterMessageReader(values)
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        return trackerController?.session?.sessionId
    }

    fun getSessionIndex(values: Map<String, Any>): Int? {
        val messageReader = GetParameterMessageReader(values)
        val trackerController = Snowplow.getTracker(messageReader.tracker)

        return trackerController?.session?.sessionIndex
    }

}

File: lib/web/snowplow_flutter_tracker_controller.dart
import 'dart:js_util';

import 'package:snowplow_flutter_tracker/web/readers/configurations.dart';
import 'package:snowplow_flutter_tracker/web/readers/messages.dart';
import 'package:snowplow_flutter_tracker/web/sp.dart';
import 'package:snowplow_flutter_tracker/web/readers/events.dart';

class SnowplowFlutterTrackerController {
  static void createTracker(ConfigurationReader configuration) {
    snowplow(
        'newTracker',
        configuration.namespace,
        configuration.networkConfig.endpoint,
        configuration.getTrackerOptions());

    if (configuration.subjectConfig != null &&
        configuration.subjectConfig?.userId != null) {
      _setUserId(configuration.namespace, configuration.subjectConfig?.userId);
    }
    addSessionContextPlugin(configuration.namespace);
  }

  static void trackStructured(EventMessageReader message) {
    trackEvent(message.structured!, message.tracker);
  }

  static void trackSelfDescribing(EventMessageReader message) {
    trackEvent(message.selfDescribing!, message.tracker);
  }

  static void trackScreenView(EventMessageReader message) {
    trackEvent(message.screenView!, message.tracker);
  }

  static void trackTiming(EventMessageReader message) {
    trackEvent(message.timing!, message.tracker);
  }

  static void trackConsentGranted(EventMessageReader message) {
    trackEvent(message.consentGranted!, message.tracker);
  }

  static void trackConsentWithdrawn(EventMessageReader message) {
    trackEvent(message.consentWithdrawn!, message.tracker);
  }

  static void trackEvent(EventReader event, String tracker) {
    snowplow('${event.endpoint()}:$tracker', jsify(event.eventData()));
  }

  static void setUserId(SetUserIdMessageReader message) {
    _setUserId(message.tracker, message.userId);
  }

  static void _setUserId(String tracker, String? userId) {
    snowplow('setUserId:$tracker', userId);
  }
}
